File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/location/engine/LocationEngineProxy.java
package org.maplibre.android.location.engine;

import android.app.PendingIntent;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import static org.maplibre.android.location.engine.Utils.checkNotNull;

public class LocationEngineProxy<T> implements LocationEngine {
  private final LocationEngineImpl<T> locationEngineImpl;
  private Map<LocationEngineCallback<LocationEngineResult>, T> listeners;

  public LocationEngineProxy(LocationEngineImpl<T> locationEngineImpl) {
    this.locationEngineImpl = locationEngineImpl;
  }

  @Override
  public void getLastLocation(@NonNull LocationEngineCallback<LocationEngineResult> callback) throws SecurityException {
    checkNotNull(callback, "callback == null");
    locationEngineImpl.getLastLocation(callback);
  }

  @Override
  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
                                     @NonNull LocationEngineCallback<LocationEngineResult> callback,
                                     @Nullable Looper looper) throws SecurityException {
    checkNotNull(request, "request == null");
    checkNotNull(callback, "callback == null");
    locationEngineImpl.requestLocationUpdates(request, getListener(callback),
      looper == null ? Looper.getMainLooper() : looper);
  }

  @Override
  public void requestLocationUpdates(@NonNull LocationEngineRequest request,
                                     PendingIntent pendingIntent) throws SecurityException {
    checkNotNull(request, "request == null");
    locationEngineImpl.requestLocationUpdates(request, pendingIntent);
  }

  @Override
  public void removeLocationUpdates(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
    checkNotNull(callback, "callback == null");
    locationEngineImpl.removeLocationUpdates(removeListener(callback));
  }

  @Override
  public void removeLocationUpdates(PendingIntent pendingIntent) {
    locationEngineImpl.removeLocationUpdates(pendingIntent);
  }

  @VisibleForTesting
  int getListenersCount() {
    return listeners != null ? listeners.size() : 0;
  }

  @VisibleForTesting
  T getListener(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
    if (listeners == null) {
      listeners = new ConcurrentHashMap<>();
    }

    T listener = listeners.get(callback);
    if (listener == null) {
      listener = locationEngineImpl.createListener(callback);
    }
    listeners.put(callback, listener);
    return listener;
  }

  @VisibleForTesting
  T removeListener(@NonNull LocationEngineCallback<LocationEngineResult> callback) {
    return listeners != null ? listeners.remove(callback) : null;
  }
}

File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/location/engine/LocationEngineProxy.kt
package org.maplibre.android.location.engine

import android.app.PendingIntent
import android.os.Looper
import androidx.annotation.VisibleForTesting
import java.util.concurrent.ConcurrentHashMap

open class LocationEngineProxy<T>(
    private val locationEngineImpl: LocationEngineImpl<T>,
) : LocationEngine {
    private var listeners: MutableMap<LocationEngineCallback<LocationEngineResult>, T>? = null

    @Throws(SecurityException::class)
    override fun getLastLocation(callback: LocationEngineCallback<LocationEngineResult>) {
        locationEngineImpl.getLastLocation(callback)
    }

    @Throws(SecurityException::class)
    override fun requestLocationUpdates(
        request: LocationEngineRequest,
        callback: LocationEngineCallback<LocationEngineResult>,
        looper: Looper?,
    ) {
        locationEngineImpl.requestLocationUpdates(request, getListener(callback), looper ?: Looper.getMainLooper())
    }

    @Throws(SecurityException::class)
    override fun requestLocationUpdates(
        request: LocationEngineRequest,
        pendingIntent: PendingIntent?,
    ) {
        locationEngineImpl.requestLocationUpdates(request, pendingIntent)
    }

    override fun removeLocationUpdates(callback: LocationEngineCallback<LocationEngineResult>) {
        removeListener(callback)?.let { locationEngineImpl.removeLocationUpdates(it) }
    }

    override fun removeLocationUpdates(pendingIntent: PendingIntent?) {
        locationEngineImpl.removeLocationUpdates(pendingIntent)
    }

    @get:VisibleForTesting
    internal val listenersCount: Int
        get() = listeners?.size ?: 0

    @VisibleForTesting
    internal fun getListener(callback: LocationEngineCallback<LocationEngineResult>): T {
        val listeners =
            this.listeners ?: ConcurrentHashMap<LocationEngineCallback<LocationEngineResult>, T>()
                .also { this.listeners = it }

        val listener = listeners[callback] ?: locationEngineImpl.createListener(callback)
        listeners[callback] = listener
        return listener
    }

    @VisibleForTesting
    internal fun removeListener(callback: LocationEngineCallback<LocationEngineResult>): T? = listeners?.remove(callback)
}
